docs/src/world_age_issue.md

```diff
@@ -0,0 +1,109 @@
+# World age issue
+
+## Motivation: lazy loading
+
+The goal of FileIO is to provide a unified IO frontend so that users can easily deal with file IO
+with the simple `load`/`save` functions. The actual IO work will be dispatched to various IO backends.
+For instance, [PNGFiles.jl](https://github.com/JuliaIO/PNGFiles.jl) is used to load PNG format images.
+If `using FileIO` loads all registered IO backends, then it would bring a significant loading latency
+to the entire ecosystem and most of them are unnecessary -- people don't do image processing probably
+don't want to load any thing related to image IO.
+
+To avoid such unnecessary loading latency, FileIO defers package loading until it's actually used.
+For instance, when you use FileIO, you'll probably observe something like this:
+
+```julia
+julia> using TestImages, FileIO
+
+julia> path = testimage("cameraman"; download_only=true)
+"/home/jc/.julia/artifacts/27a4c26bcdd47eb717bee089ec231a899cb8ef69/cameraman.tif"
+
+julia> load(path) # actual backend loading happens here
+[ Info: Precompiling ImageIO [82e4d734-157c-48bb-816b-45c225c6df19]
+[ Info: Precompiling TiffImages [731e570b-9d59-4bfa-96dc-6df516fadf69]
+...
+```
+
+## The hidden issue
+
+Although this lazy loading trick solves the large loading latency issue, it isn't a normal practice
+in Julia because it introduces a so-called _world age issue_ or _world age problem_. The world age
+issue happens when the compiled function becomes outdated.
+
+Now, let's try something new to uncover this issue:
+
+```julia
+julia> using IndirectArrays, ImageCore
+
+julia> img = IndirectArray(rand(1:5, 4, 4), rand(RGB, 5))
+4×4 IndirectArray{RGB{Float64}, 2, Int64, Matrix{Int64}, Vector{RGB{Float64}}}:
+...
+
+julia> save("indexed_image.png", img)
+```
+
+and **reopen a new julia REPL** and wrap the `load` into a function, you'll observe something like
+this:
+
+```julia
+julia> using FileIO
+
+julia> f() = length(load("indexed_image.png"))
```

Comment:
```suggestion
julia> f() = size(load("indexed_image.png"))
```
Some people may not quickly see the connection between `length` and the error about `size`.